---
title: "Project_Guide"
author: "Kyungjin Sohn"
date: '2021 6 11 '
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This document is a simple guide about the "Comparative robustness analysis" project. To execute all the code used in this project, the code must be in the correct directory. Please check the link before executing each code. Before entering a description, this project was developed in the following environment.

    - R : 4.1.0
    - RStudio : 1.4.1717
    - OS : Windows10    
    

## 1. Calculate robustness scores

First, we need to compute robustness scores using cDNA data. The data is in the "./data" folder and contains six types of cDNA data; HomoSapiens, MusMusculus, Dmelanogaster, Celegans, Scerevisiae, and Ecoli.

Run the first code named *main_01.R*. This function takes arguments and starts the calculation. If you need more description, run the following command on your server.

```{r eval=FALSE}
Rscript main_01.R -h
```

And the sample command to execute the main_01.R function is like:

```{r eval=FALSE}
nice -n19 Rscript main_01.R -i Scerevisiae.fasta.gz -o Scerevisiae -s Scerevisiae
```


## 2. Categorize score data 

To do an in-depth analysis of the score data, we categorized the score data based on gene ontology. (There are three types of gene ontology: BP(=Biological Process), CC(Cellular Component), and MF(Molecular Function)) The methods for matching Go terms and ontologies with score data are species-specific, so we distinguished them by different codes. Besides, we used Nucleotide ID to classify both HomoSapiens and MusMusculus score data, so we used the same code for data classification of both species. The function names for this part all begin with *cateBonto_(species).R*
```{r eval=FALSE}
cateBonto_NucleotideID("HomoSapiens", "Homo Sapiens")
cateBonto_Ecoli()
```

Since not all sequences have matching GO terms and ontologies(depending on the database), there will be several sequences that do not belong to any categorized score data. There is a code that recombines categorized data. The result will be score data containing only sequences with matching GO terms and ontologies.
```{r eval=FALSE}
combine_ONTO_data("HomoSapiens")
```


## 3. Create random sequence

There are two ways to create a random sequences. First, we can maintain the original GC-contents and make the rest completely random. On the other hand, we can also create data that maintain amino acids and convert codons into only other synonymous codon. We will maintain the original length of the sequence in both ways.

#### 1) Method 1\
The GC-contents are the same as the original data. (The function takes two arguments; a species name and a seed number.)
```{r eval=FALSE}
createRandomSeq_GC("Scerevisiae", 1111)
```

#### 2) Method 2\
The composition of amino acids is the same as the original data. (The function takes two arguments; species name and seed number.)
```{r eval=FALSE}
createRandomSeq_AA("Scerevisiae", 1111)
```


## 4. Draw histogram by GO term

#### 1) Compute robustness score means for GO term\
Before drawing a histogram, we must select the GO term(s) to be the main sample. Using the categorized data based on ontology, we can calculate the robustness score average for each GO group. According to the law of large numbers, we used only GO terms with more than 30 sequences. The result is saved in a CSV file.
```{r eval=FALSE}
scoremeanBGO("HomoSapiens", "BP")
```

To run this function  *scoremeanBGO.R*, another function is also needed.

        - subNmodi_scoreONTO.R : 
          subset from the categorized score data and modify score to standardize them
        
#### 2) Draw histogram\
The function *draw_histoBGO.R* draws histograms when you pass the species and ontology arguments. This function draws three types of histograms: using the CSV file calculated above, it draws a histogram of all GOs in the top 10%, all GOs in the bottom 10%, and each GO in the top or bottom 10%. In addition, all of these histograms include multiple distributions of random, selected, and total distributions.\

The function *draw_histoBGO.R* requires many other functions.

        - subNmodi_scoreONTO.R : (described above)
        - subNmodi_score.R : subset from the score data and modify score to standardize them
        - sortGO.R : find GOs ranked in the top or bottom 10% 
        - plot_histoBGO.R : draw a histogram according to GO terms and save it as a png file
                            [*Need to change seed number and random data type if necessary]

## 5. Draw histogram by gene
The function *draw_histoBgene.R* draws histograms when the species arguments is passed as an argument. This function draws two types of histograms: the histogram includes the score distribution for a particular species and the score distribution for random data generated according to the 3-1 or 3-2 method. A special part of the function is that among random score data of the same type, there is a subfunction called "*combine_random_data.R*" that combines data with different seed numbers. Thus, it is necessary to change the seed vector based on the seed numbers used when creating random data.\

The required functions to run *draw_histoBgene.R* function.

        - subNmodi_score.R : (described above)
        - plot_histoBgene.R : 
          draw a histogram according to gene of the species and save it as a png file
        - combine_random_data.R : 
          combines data with different seed numbers among random score data of the same type
    